source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,259 @@
+#include "common/network/apple_dns_impl.h"
+
+#include <dns_sd.h>
+
+#include <chrono>
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <string>
+
+#include "envoy/common/platform.h"
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Network {
+
+AppleDnsResolverImpl::AppleDnsResolverImpl(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher) {
+  ENVOY_LOG(warn, "Constructing DNS resolver");
+  initializeMainSdRef();
+}
+
+AppleDnsResolverImpl::~AppleDnsResolverImpl() {
+  ENVOY_LOG(warn, "Destructing DNS resolver");
+  deallocateMainSdRef();
+}
+
+void AppleDnsResolverImpl::deallocateMainSdRef() {
+  ENVOY_LOG(warn, "DNSServiceRefDeallocate main sd ref");
+  // dns_sd.h says:
+  //   If the reference's underlying socket is used in a run loop or select() call, it should
+  //   be removed BEFORE DNSServiceRefDeallocate() is called, as this function closes the
+  //   reference's socket.
+  sd_ref_event_.reset();
+  DNSServiceRefDeallocate(main_sd_ref_);
+}
+
+void AppleDnsResolverImpl::initializeMainSdRef() {
+  auto error = DNSServiceCreateConnection(&main_sd_ref_);
+  if (error) {
+    throw EnvoyException("Error in DNSServiceCreateConnection");
+  }
+
+  auto fd = DNSServiceRefSockFD(main_sd_ref_);
+  if (fd == -1) {
+    throw EnvoyException("Error in DNSServiceRefSockFD");
+  }
+  ENVOY_LOG(warn, "DNS resolver has fd={}", fd);
+
+  sd_ref_event_ = dispatcher_.createFileEvent(
+      fd, [this](uint32_t events) { onEventCallback(events); }, Event::FileTriggerType::Level,
+      Event::FileReadyType::Read);
+  sd_ref_event_->setEnabled(Event::FileReadyType::Read);
+}
+
+void AppleDnsResolverImpl::onEventCallback(uint32_t events) {
+  ENVOY_LOG(warn, "DNS resolver file event");
+  if (events & Event::FileReadyType::Read) {
+    DNSServiceProcessResult(main_sd_ref_);
+  }
+}
+
+ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
+                                              DnsLookupFamily dns_lookup_family,
+                                              ResolveCb callback) {
+  ENVOY_LOG(warn, "DNS resolver resolve={}", dns_name);
+  std::unique_ptr<PendingResolution> pending_resolution(
+      new PendingResolution(*this, callback, dispatcher_, main_sd_ref_, dns_name));
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    return nullptr;
+  }
+
+  // If the query was synchronously resolved, there is no need to return the query.
+  // FIXME: I don't think synchronous resolution is possible from my read of dns_sd.h,
+  // and tests with local resolution.
```

Comment:
I'm confused by this also. Is this referring to whether a query was completed as part of an (unrelated) batch? Or if it actually completed inline and synchronously. I guess a cache hit is a plausible explanation.

Reply:
The latter.

Exactly re: cache. However, I have not been able to trigger a synchronous resolution. My thought was that I was going to code defensively just in case this is possible.